add tensorboard Histogram

cjcchen opened the pull request:
new pr

ryanjulian committed:
TravisCI support

ryanjulian committed:
Protect StatefulPool from class methods

StatefulPool doesn't support batching over with class methods because it
always passes `G` as the first argument to the worker function. If one of the
`run_` methods in StatefulPool is called with a class method it can lead to
a silent lock-up of the pool, which is very difficult to debug.

Note: this bug does not appear unless n_parallel > 1

ryanjulian committed:
[travisci] Correct changed file detection

cjcchen committed:
Add TensorBoard Support

Adds TensorBoard support for basic key-value pairs. Anything logged via
`logger.record_tabular()` is also available via TensorBoard.

cjcchen committed:
Add support for dm_control environments (#48)

Adds a support for training against [dm_contro](https://github.com/deepmind/dm_control) environments, and an associated viewer. 

dm_control doesn't support direct 3D rendering yet (only pixels), so we feed the pixels to a pygame window.

cjcchen committed:
add histogram

cjcchen committed:
add histogram

cjcchen committed:
move example files

cjcchen committed:
move example files

Commit:
Add TensorFlow GPU v1.8 during the setup of the virtual env (#9) (#59)

The selection of the regular TensorFlow package or GPU-
accelerated TensorFlow was added into the setup_linux
script.
Although no error is produced if both packages are installed,
the GPU package won't work if the regular package is installed
as well. Therefore, both packages were removed from the conda
environment file.
In the setup_linux script, the selection of the TF package is
done after the environment is created, leaving the regular
release as the default package. Two guards were added for the
creation and update of the conda environment to verify if their
execution was successful or else exit the script with an error
message.
The TF version was set to 1.8 which is the most recent release
that works for both the GPU and regular version.
As a side note, the prettytensor package removed from the conda
environment has the regular TF package as a dependency, and
prettytensor is currently not used in the rllab project, so
it was replaced by the selection of the regular TF package
in the setup script.

ryanjulian commented:
Related to #54

Commit:
Cleanup list of dependencies in conda environment (#56) (#60)

It was verified that each of the dependencies removed were not
imported or had any package dependants in rllab.
The new environment file was tested with a fresh copy of the
rllab project by running the setup_linux script, and with the
training of the swimmer in Theano and cartpole in TensorFlow
with plot=True.
The channels of the removed packages were discarded as well.

ryanjulian commented on rllab/misc/logger.py:
```diff
@@ -1,12 +1,17 @@
-import os
+
 import os.path as osp
+import os
```
Please reorder imports to conform to PEP8.

https://www.python.org/dev/peps/pep-0008/#imports

ryanjulian commented on rllab/misc/tensorboard_summary.py:
```diff
@@ -0,0 +1,142 @@
+from rllab.misc.console import mkdir_p
```
Please reorder imports to conform to PEP8

https://www.python.org/dev/peps/pep-0008/#imports

cjcchen replied:
It seems the order is fine now.

cjcchen replied:
I have changed the order. System library goes first, the second one is the third-party library, and the last one is our library.

>import os
>import tensorflow as tf
>from rllab.misc.console import mkdir_p

ryanjulian replied:
These are ordered correctly, but need to be separated into groups:

> Imports are always put at the top of the file, just after any module comments and docstrings, and before module globals and constants.
> 
> Imports should be grouped in the following order:
> 
> Standard library imports.
> Related third party imports.
> Local application/library specific imports.
>
> __You should put a blank line between each group of imports.__

```python
import os

import tensorflow as tf

from rllab.misc.console import mkdir_p
```

ryanjulian reviewed (changes requested):
This looks good, it just needs some polish to conform to PEP8.

In your last commit, can you please include a commit message conforming to the commit message guidelines?
https://chris.beams.io/posts/git-commit/
https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

ryanjulian commented:
I'll approve once the comments are addressed and the CI is passing.

CatherineSue committed:
Support std_share_network in all GaussianMLP* classes (#69)

Implement std_share_network architecture in the GaussianMLP*
classes of TensorFlow.

The std_share_network creates a single neural net with output
length of 2 * action_dimension. The first half output params
are the means params, and the second half params are the
log_std params.

As a result, the GaussianMLP* classes can support a new
architecture.

Refer to: #44

Commit:
Import classes and modules in __init__.py of each package (#66)

Imports classes and modules in the __init__.py of each package to make
package-level imports shorter.

Commit:
Fix error in Theano when using GPU and TF

The configuration of the GPU device in Theano is set if USE_TF is false and USE_GPU is true. Since USE_TF is only set to True when using AWS EC2, it's now also set as a parameter in the run_experiment_list function.
To avoid the conflict with Theano when using TF and GPU, set both use_tf and use_gpu to True when calling run_experiment_list.
To enable and disable GPU with TensorFlow, an if statement was inserted to make the GPU invisible when use_gpu=False.
Also, an environment variable for use_tf for non-local modes such as local_docker or EC2 was added if it's required to switch between Theano and TensorFlow.
Finally, the corresponding documentation was added in experiments.rst and the script instrument.py was formatted with PEP8 using YAPF.

hejia-zhang committed:
add gazebo sawyer (#49)

Design the ros environment support for rllab. Add sawyer simulation support and gazebo environment support.

Refer to: #49

Commit:
Remove convenience imports that generate circular dependencies (#87)

When the convenience imports were introduced, they accidentally created
a circular dependency was  between packages.

To avoid future crashes, an analysis of the code imports was performed
to find all the circular dependencies in rllab. We found no no circular
dependencies in the TensorFlow tree, or between packages in the
TensorFlow and Theano tree. There were circulater dependencies in the
Theano tree.

Circular dependencies in the Theano tree are the following:
- The package misc with algos, baselines, core, and viskit.
- The package algos with sampler

Since misc shares so many circular dependencies, conflicting imports
were removed from its file __init__.py.

Regarding algos and sampler, the issue is generated by imports in
sampler in its file __init__.py, so they were removed as well.

Due to these removals, some imports had to be restored to its previous
long package path.

cjcchen committed:
reorder imports in logger.py

cjcchen committed:
reorder imports in logger.py

cjcchen committed:
reorder imports in tensorboard_summary.py

cjcchen committed:
reorder imports in tensorboard_summary.py

cjcchen committed:
Merge branch 'integration' into histogram

cjcchen committed:
reformat tensorboard_summary.py in yapf

cjcchen committed:
reformat tensorboard_summary.py in yapf

cjcchen committed:
Merge branch 'histogram' of https://github.com/cjcchen/rllab into histogram

cjcchen committed:
Upgrade Theano to 1.0.1 (#65)

Upgrades Theano to 1.0.1

Commit:
Add variable scope to symbolic operations using TensorFlow (#72)

Add variable scope to symbolic operations using TensorFlow

A variable scope facilitates the reading of a TensorFlow graph by
grouping tensor objects in hierarchies.

In rllab, the hierarchies are defined with primitive objects and
symbolic operations, where the latter is a member function of a
primitive.

In this change, the primitives are the algorithms, networks,
distributions, optimizers, policies, Q-functions and regressors.

An example of a primitive is the DiagonalGaussian class, which
implements a probability distribution, and the symbolic function is
kl_sym, which implements the Kullback–Leibler divergence. The idea of
implementing the variable scope is that all the tensor operations in
kl_sym are encapsulated within the hierarchy DiagonalGaussian/kl_sym.

Each primitive and symbolic function have the parameter "name", which
has a default value equal to the primitive or symbolic function name,
but developers can set those parameters as they may find more
convenient, so the previous example could be changed to be:
distribution/divergence.

A context class was added to the file tensor_utils.py to verify if the
variable scope of the corresponding primitive is already set, in order
to set it in case it's not, and remove it once the symbolic function is
exiting.

The only caveat with variable scopes is that they work based on the call
stack and not with a class or file scope, so even if the same scope is
used in two different call stacks, TensorFlow adds an index to the
primitive or symbolic function names to make each scope unique.
Therefore, not all symbolic operations performed in a primitive will
appear under the same primitive scope, but they will be split in
different instances of the primitive scope (e.g. DiagonalGaussian and
DiagonalGaussian_1).

CatherineSue committed:
Support for tensor logging to tensorboard (#88)

Add a customized tensor scalar to tensorboard by using the
custom_scalar plugin in tensorboard. Each line in the scalar
corresponds to an element in the tensor.

Wrap the tensorboard logging module into a new class `Summary`
in file rllab/misc/tensor_summary.py. It supports both the
simple value and tensor logging. It also saves the
computation graph created by rllab.

To record the tensor into tensorboard, use the
`record_tensor` function in file rllab/misc/logger.py.

Refer to: #39, #38

CatherineSue committed:
Fix missing imports and other small errors (#99)

1. Add missing imports and optimize imports.
2. Delete duplicate parameter name.
3. Delete duplicate parameter initialization.

cjcchen committed:
Merge branch 'integration' into histogram

cjcchen committed:
yapf format for logger.py

cjcchen committed:
yapf format for logger.py

ryanjulian commented on examples/example_tensorboard_logger.py:
```diff
@@ -0,0 +1,17 @@
+from rllab.misc import logger
```
PEP8: import order

cjcchen committed:
rebase from integration

CatherineSue commented on rllab/misc/logger.py:
```diff
@@ -40,8 +45,8 @@
 
-_tensorboard_default_step = 0
 _tensorboard_step_key = None
+_tensorboard = Summary()
```
This seems to be a conflict with `_tensorboard = TensorBoardOutput()` in line 51. Is it possible to merge class Summary with TensorBoardOutput?

cjcchen committed:
reorder the imports

CatherineSue commented on rllab/misc/logger.py:
```diff
@@ -154,16 +153,17 @@ def log(s, with_prefix=True, with_timestamp=True, color=None):
 
 def record_tabular(key, val):
-    _tensorboard.record_scale(str(key), val)
+    _tensorboard_summary.record_scale(str(key), val)
```
I think there is a merge error here. In this commit, there should be a record_tensor() function in line 156. https://github.com/CatherineSue/rllab/blob/9530e7f6e2687fe55355c78c790a93849b7eb4ba/rllab/misc/logger.py

cjcchen replied:
Why does it not show that I remove them?
And I found some features are the same in our codes. I think you could merge my code to yours and remove mine.

CatherineSue replied:
> Why does it not show that I remove them?

It shows here: https://github.com/ryanjulian/rllab/pull/58/commits/fa6ac5ec47e2f51ede79b434f1bf30bea53fc9ca
But it shouldn't be removed.

> And I found some features are the same in our codes. I think you could merge my code to yours and remove mine.

Yeah, this could work. I will approve this and merge your code.

cjcchen committed:
merge tensorboard output and tensorboard summary

cjcchen committed:
when dumping scale, it uses tensorboard output, when dumping histogram, it uses tensorboard summary

cjcchen committed:
Merge branch 'integration' into histogram
